### PR TITLE
test(annotation): cover distribution/main.py

### PR DESCRIPTION
**Pull request by UkiStefanovic**
This Pull request adds unit tests for multiple functions from the distribution/main.py module. Functions covered with this pull request are:
-  choose_validators_users, 
- prepare_users, 
- find_users_share_loads, 
- find_equal_files, 
- find_small_files, 
- prepare_response.

You can find the unit tests for these function in /tests/test_distribution.py

### CODE REVIEW

File: annotation/tests/test_distribution.py [hunk 1]
```diff
@@ -1277,3 +1283,191 @@ def test_get_page_number_combinations(
     expected: List[int],
 ) -> None:
     assert get_page_number_combinations(file_pages, user_pages) == expected
+
+
+# ADDITIONAL TESTS
+
```
Comment (Negoje) [inline annotation/tests/test_distribution.py:1289]: We should check if similar mock objects already exist, so that we do not create unnecessary new objects.

**Review by Negoje — COMMENTED** [review]
All tests are green. Good job!

File: annotation/.pre-commit-config.yaml [hunk 2]
```diff
@@ -1,7 +1,10 @@
+# https://pre-commit.com
+# https://pre-commit.com/hooks.html
+
```
Comment (cakeinsauce) [inline annotation/.pre-commit-config.yaml:3]: No need for that

File: annotation/tests/test_distribution.py [hunk 3]
```diff
@@ -1277,3 +1283,191 @@ def test_get_page_number_combinations(
     expected: List[int],
 ) -> None:
     assert get_page_number_combinations(file_pages, user_pages) == expected
+
+
+# ADDITIONAL TESTS
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1288]: No need for this comment

File: annotation/tests/test_distribution.py [hunk 4]
```diff
@@ -1277,3 +1283,191 @@ def test_get_page_number_combinations(
...
+    {
+        "user_id": "405ef0e2-b53e-4c18-bf08-c0871615d99d",
+        "default_load": 100,
+        "overall_load": 0,
+    },
+    {
+        "user_id": "ba3e0ccf-8661-4ed3-892b-7c291160f631",
+        "default_load": 96,
+        "overall_load": 0,
+    },
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1300]: Do you need so much users for your tests? Usually you would have something normal and edge cases, apply this principle here as well. And make it a fixture, please. Same applies to ANNOTATION_TASKS_CHOOSE_VALIDATORS

File: annotation/tests/test_distribution.py [hunk 5]
```diff
@@ -1277,3 +1283,191 @@ def test_get_page_number_combinations(
...
+            "user_id": "c080408e-0077-4c20-b520-bd4b1541ca59",
+            "default_load": 0,
+            "overall_load": 50,
+        },
+    ],
+    [
+        {
+            "user_id": "6e2ce9ac-2fe8-4cc7-bdcd-bac90faa9247",
+            "default_load": 95,
+            "overall_load": 0,
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1390]: Expected output is always related to a certain test, no need to keep in in a module level. Additionally, all data you prepared above should be used in test directly if you dont utilize in multiple places

File: annotation/tests/test_distribution.py [hunk 6]
```diff
@@ -1277,3 +1283,191 @@ def test_get_page_number_combinations(
...
+    "annotators,"
+    "validators,"
+    "annotation_tasks,"
+    "expected_output",
+    [
+        (
+            False,
+            ValidationSchema.cross,
+            USERS[0:2] + USERS[7:],
+            USERS[2:7],
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1428]: This is not really readable. After you change testing data according to my comments above, change this as well. And do you really need so much users to test the functionally? Take a bare minimum to cover different cases

File: annotation/tests/test_distribution.py [hunk 7]
```diff
@@ -1277,3 +1283,191 @@ def test_get_page_number_combinations(
...
+            True,
+            ValidationSchema.validation_only,
+            USERS[0:2] + USERS[7:],
+            USERS[2:7],
+            ANNOTATION_TASKS_CHOOSE_VALIDATORS,
+            EXPECTED_OUTPUT[1],
+        ),
+    ],
+)
+@pytest.mark.unittest
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1450]: No need for this

File: annotation/tests/test_distribution.py [hunk 8]
```diff
@@ -1277,3 +1283,191 @@ def test_get_page_number_combinations(
...
+def test_choose_validators_users(
+    split_multipage_doc_setup,
+    validation_type,
+    annotators,
+    validators,
+    annotation_tasks,
+    expected_output,
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1457]: Add type hints, please

File: annotation/tests/test_distribution.py [hunk 9]
```diff
@@ -1277,3 +1283,191 @@ def test_get_page_number_combinations(
...
+):
+    # SETUP
+    # Set environment variable SPLIT_MULTIPAGE_DOC
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1460]: No need for such comments, code is pretty explanatory. Whenever you think about writing a new comment, ask yourself - is my code understandable without comments? If not, most likely you should change a code

File: annotation/tests/test_distribution.py [hunk 10]
```diff
@@ -1277,3 +1283,191 @@ def test_get_page_number_combinations(
...
+    if split_multipage_doc_setup:
+        os.environ["SPLIT_MULTIPAGE_DOC"] = "1"
+    else:
+        del os.environ["SPLIT_MULTIPAGE_DOC"]
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1464]: You should never mock env vars like that (any mutable object). It brings a lot of side effects, you don't want to mess with that later.
Check out this
https://docs.python.org/3/library/unittest.mock.html#patch-dict

File: annotation/tests/test_distribution.py [hunk 11]
```diff
@@ -1277,3 +1283,191 @@ def test_get_page_number_combinations(
...
+
+    output = choose_validators_users(
+        validation_type, annotators, validators, annotation_tasks
+    )
+
+    # Checking is lists have the same elements
+    # expected_output is already sorted by 'user_id'
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1471]: No need for this comment

File: annotation/tests/test_distribution.py [hunk 12]
```diff
@@ -1277,3 +1283,191 @@ def test_get_page_number_combinations(
...
+    sorted_output = sorted(output, key=lambda x: x["user_id"])
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1472]: Could you elaborate why we sort output? Interfere with output for testing is not the best practice. Add logic to a test overall

Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1472]: Yes, I've noticed that the function returns the list of same elements, but it shuffles them every time. So my idea was to sort the output list because then sorted_output would always be the same list for every test run, but output wouldn't. Ideally, we would like the function to return a set, and then I wouldn't have sorted the output, I would have just done assert output==expected, but because output list is mixed every run, I sorted it. I'm currently changing the test cases to be very simple and possibly will not have sorting in the teardown. Should I never sort the output if I just need to check if elements of two lists are the same? And if so, why?

Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py]: Nevermind, this code is now irrelevant. Check out my fixed test

File: annotation/tests/test_distribution.py [hunk 13]
```diff
@@ -1277,3 +1284,48 @@ def test_get_page_number_combinations(
...
+        ),
+        (
+            "1",
+            ValidationSchema.hierarchical,
+            [{"user_id": "0"}, {"user_id": "1"}],
+            [{"user_id": "2"}],
+            [{"user_id": "1"}],
+            [{"user_id": "2"}],
+        ),
+    ],
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1315]: list -> tuple (outer one)

File: annotation/tests/test_distribution.py [hunk 14]
```diff
@@ -1277,3 +1284,48 @@ def test_get_page_number_combinations(
...
+def test_choose_validators_users(
+    split_multipage_doc_setup: str,
+    validation_type: ValidationSchema,
+    annotators: List[User],
+    validators: List[User],
+    annotation_tasks: List[Task],
+    expected_output: List[User],
+):
+    with patch.dict(
+        "os.environ", {"SPLIT_MULTIPAGE_DOC": split_multipage_doc_setup}
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1326]: Patching os.environ is not the best practice. Patching env vars overall. Let's patch annotation.distribution.main.SPLIT_MULTIPAGE_DOC value itself

**Review by cakeinsauce — APPROVED** [review]
LGTM!

File: annotation/tests/test_distribution.py [hunk 15]
```diff
@@ -1277,3 +1286,97 @@ def test_get_page_number_combinations(
...
+        ),
+    ),
+)
+def test_choose_validators_users(
+    split_multipage_doc_setup: str,
+    validation_type: ValidationSchema,
+    annotators: List[User],
+    validators: List[User],
+    annotation_tasks: List[Task],
+    expected_output: List[User],
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1325]: You pass List of dicts, not User or Task

File: annotation/tests/test_distribution.py [hunk 16]
```diff
@@ -1277,3 +1286,97 @@ def test_get_page_number_combinations(
...
+    with patch(
+        "annotation.distribution.main.read_user",
+        side_effect=read_user_side_effect,
+    ), patch(
+        "annotation.distribution.main.create_user",
+        side_effect=create_user_side_effect,
+    ):
+        output = prepare_users(db, users_id)
+        for a, e in zip(output, expected_output):
+            assert a.user_id == e.user_id
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1382]: Why dont we compare output with expected output? When we have such asserts, if they fail, it is harder to read and pinpoint what exactly failed and why

File: annotation/tests/test_distribution.py (hunk 15 above)
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1325]: I thought that was okay cause I saw Task type defined as:
Task = Dict[str, Union[int, List[int], bool]]

File: annotation/tests/test_distribution.py (hunk 16 above)
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1382]: The User class does not have __eq__ and so I had to check their id's manually. Is there a better way to check output==expected?s

File: annotation/tests/test_distribution.py (hunk 15 above)
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1325]: Also, Users are for some reason unsubscriptable and you cant call user['user_id'] on a User object, so I went with dictionaries for Users as thats probably how they are stored in the database

File: annotation/tests/test_distribution.py (hunk 16 above)
Comment (cakeinsauce) [reply annotation/tests/test_distribution.py:1387]: Ask Djordje, he had the same issue, he will help

File: annotation/tests/test_distribution.py [hunk 17]
```diff
@@ -1277,3 +1291,238 @@ def test_get_page_number_combinations(
...
+        print("CREATE ME")
+        return User(user_id=user_id)
+
+    with patch(
+        "annotation.distribution.main.read_user",
+        side_effect=read_user_side_effect,
+    ), patch(
+        "annotation.distribution.main.create_user",
+        side_effect=create_user_side_effect,
+    ):
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1384]: Please, adjust your patches with side effects, you can make it much simplier without creation of these functions.

File: annotation/tests/test_distribution.py [hunk 18]
```diff
@@ -1277,3 +1291,238 @@ def test_get_page_number_combinations(
...
+    ),
+    (
+        (
+            [{"user_id": "0", "overall_load": 10, "default_load": 10}],
+            10,  # all_job_pages_sum
+            10.0,  # average_job_pages
+            10,  # users_default_load
+            10,  # users_overall_load
+            1.0,  # expected_output
+            [1],  # expected_share_loads
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1408]: you have one parametrized case, no need to use it then

File: annotation/tests/test_distribution.py [hunk 19]
```diff
@@ -1277,3 +1291,238 @@ def test_get_page_number_combinations(
...
+    output = find_users_share_loads(
+        users,
+        all_job_pages_sum,
+        average_job_pages,
+        users_default_load,
+        users_overall_load,
+    )
+    assert output == expected_output
+    for user, expected_share_load in zip(users, expected_share_loads):
+        assert user["share_load"] == expected_share_load
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1430]: dont use such constructions in tests, please, let's make it simplier

File: annotation/tests/test_distribution.py [hunk 20]
```diff
@@ -1277,3 +1291,238 @@ def test_get_page_number_combinations(
...
+                {"pages_number": 30, "file_id": 2},
+                {"pages_number": 40, "file_id": 3},
+            ],
+            90,
+            [
+                {"pages_number": 20, "file_id": 1},
+                {"pages_number": 30, "file_id": 2},
+                {"pages_number": 40, "file_id": 3},
+            ],
+        ),
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1454]: you only change the actual number, let's simplify parametrize cases and construct full input/output data in the test

File: annotation/tests/test_distribution.py [hunk 21]
```diff
@@ -1277,3 +1291,238 @@ def test_get_page_number_combinations(
...
+        (
+            [
+                {"pages_number": 10, "file_id": 0},
+                {"pages_number": 30, "file_id": 1},
+                {"pages_number": 0, "file_id": 2},
+            ],
+            20,
+            "a",
+            ([{"pages_number": 10, "file_id": 0}], 0),
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1482]: same as previous comment

File: annotation/tests/test_distribution.py [hunk 22]
```diff
@@ -1277,3 +1291,238 @@ def test_get_page_number_combinations(
...
+def test_find_equal_files(
+    files: List[Dict[str, int]],
+    user_pages: int,
+    expected_output: List[Dict[str, int]],
+):
+    assert find_equal_files(files, user_pages) == expected_output
+
+
+@pytest.mark.parametrize(
+    ("files", "user_pages", "SPLIT_MULTIPAGE_DOC_setup", "expected_output"),
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1466]: format SPLIT_MULTIPAGE_DOC_setup to follow pep8

File: annotation/tests/test_distribution.py [hunk 23]
```diff
@@ -1277,3 +1291,238 @@ def test_get_page_number_combinations(
...
+    user_pages: int,
+    SPLIT_MULTIPAGE_DOC_setup: str,
+    expected_output,
+):
+    def find_files_for_task_side_effect(files, pages_for_task):
+        lst = []
+        for file in files:
+            if file["pages_number"] and file["pages_number"] in pages_for_task:
+                lst.append(file)
+        return lst
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1497]: too complicated. if you do that for your patch, it is most likely due to incorrect patching. think what you could patch to avoid such complex thing

File: annotation/tests/test_distribution.py [hunk 24]
```diff
@@ -1277,3 +1291,238 @@ def test_get_page_number_combinations(
...
+                mock_arg,
+                mock_arg,
+                mock_arg,
+                mock_arg,
+                mock_arg,
+                mock_arg,
+                mock_arg,
+            )
+            == expected_output
+        )
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1528]: what do we test here? seems like mock testing

File: annotation/tests/test_distribution.py [hunk 25]
```diff
@@ -1277,3 +1291,238 @@ def test_get_page_number_combinations(
...
+            ["b908ccf0-6eab-4b8e-afb7-913372ddf4e5"],
+            [
+                User(user_id="b908ccf0-6eab-4b8e-afb7-913372ddf4e5"),
+                User(user_id="00d37dc4-b7da-45e5-bcb5-c9b82965351c"),
+            ],
+        ),
+    ),
+)
+def test_prepare_users(
+    users_id: List[UUID], example_db: Tuple[UUID], expected_output: List[User]
```
[the remark this answers is not in the document]
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1364]: The only error mypy is telling me is: Unexpected keyword argument "user_id" for "User". Im not sure how else to create User objects with a given user_id. Should I change User objects into dictionaries or should I do something else?

File: annotation/tests/test_distribution.py (hunk 17 above)
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1384]: Maybe instead of the function read_user_side_effect and create_user_side_effect, I could just pass lambdas. Were you thinking of some other way?

File: annotation/tests/test_distribution.py (hunk 24 above)
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1530]: The function prepare_response only does:
return [ task for task in distribute(...)]
I didnt really understand why we have that function, since it returns the same list like distribute(...)
I only made this test to cover the function, but it really doesnt test any functionality

File: annotation/annotation/models.py [hunk 26]
```diff
@@ -200,6 +200,17 @@ class User(Base):
...
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, User):
+            return NotImplemented
+
+        for column in self.__table__.columns:
+            if column.key != "_sa_instance_state" and other.__getattribute__(
+                column.key
+            ) != self.__getattribute__(column.key):
+                return False
+        return True
```
Comment (UkiStefanovic) [inline annotation/annotation/models.py:212]: After adding this __eq__, now i can test:
assert user1 == user2

File: annotation/tests/test_distribution.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1366]: Yesterday, I fixed the data types

File: annotation/tests/test_distribution.py (hunk 18 above)
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1402]: To make sure all parts of code in the function are tested, I'll add one more test case

File: annotation/tests/test_distribution.py (hunk 23 above)
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1499]: Now that I think about it, we don't need to patch the function _find_files_for_task_. I will remove this patch and it will make the test much simpler

Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1499]: Also, I've just now realized that my test doesn't cover one edge case. I will add another test case for this test

File: annotation/tests/test_distribution.py (hunk 20 above)
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1476]: I noticed that I needed to add another test case. Please tell me if you think I should still simplify the parametrize. The new test case makes it a bit more complicated

File: annotation/annotation/models.py [hunk 27]
```diff
@@ -200,6 +200,17 @@ class User(Base):
         "Job", secondary=association_job_owner, back_populates="owners"
     )
 
+    def __eq__(self, other) -> bool:
```
Comment (cakeinsauce) [inline annotation/annotation/models.py:203]: add typing for arg please

File: annotation/tests/test_distribution.py [hunk 28]
```diff
@@ -1277,3 +1291,278 @@ def test_get_page_number_combinations(
...
+        "annotation_tasks",
+        "expected_output",
+    ),
+    (
+        (
+            "",
+            ValidationSchema.cross,
+            [{"user_id": "0"}, {"user_id": "1"}],
+            [{"user_id": "0"}, {"user_id": "1"}],
+            [],
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1310]: you can simply specify tuple of ids and create list of dicts with user ids in the actual test, effectively making this parametrize much smaller and easier to read

File: annotation/tests/test_distribution.py [hunk 29]
```diff
@@ -1277,3 +1291,278 @@ def test_get_page_number_combinations(
...
+)
+def test_prepare_users(
+    users_id: List[UUID], example_db: Tuple[UUID], expected_output: List[User]
+):
+    db = Mock()
+
+    with patch(
+        "annotation.distribution.main.read_user",
+        side_effect=lambda db, user_id: User(user_id=user_id)
+        if user_id in example_db
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1371]: in this test you dont have any case where user_id is not present in example_db. Please adjust test or better add edge cases

File: annotation/tests/test_distribution.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (cakeinsauce) [reply annotation/tests/test_distribution.py:1364]: `Tuple[UUID, ...]`

I believe I have already shared it, but once again
[link](https://docs.python.org/3/library/typing.html#:~:text=To%20denote%20a%20tuple%20which%20could%20be%20of%20any%20length%2C%20and%20in%20which%20all%20elements%20are%20of%20the%20same%20type%20T%2C%20use%20tuple%5BT%2C%20...%5D.%20To%20denote%20an%20empty%20tuple%2C%20use%20tuple%5B()%5D.%20Using%20plain%20tuple%20as%20an%20annotation%20is%20equivalent%20to%20using%20tuple%5BAny%2C%20...%5D%3A)

File: annotation/tests/test_distribution.py [hunk 30]
```diff
@@ -1277,3 +1291,278 @@ def test_get_page_number_combinations(
...
+        "expected_users",
+    ),
+    (
+        (
+            [{"user_id": "0", "overall_load": 10, "default_load": 10}],
+            10,  # all_job_pages_sum
+            10.0,  # average_job_pages
+            10,  # users_default_load
+            10,  # users_overall_load
+            1.0,  # expected_output
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1398]: you dont need such comments for params you have already denoted

Comment (cakeinsauce) [reply annotation/tests/test_distribution.py:1398]: Same applies for comments below

File: annotation/tests/test_distribution.py (hunk 20 above)
Comment (cakeinsauce) [reply annotation/tests/test_distribution.py:1476]: I did not get the problem here. You can only provide numbers, it wont complicate the test

File: annotation/tests/test_distribution.py [hunk 31]
```diff
@@ -1277,3 +1291,278 @@ def test_get_page_number_combinations(
...
+            "",
+            ([], 0),
+        ),
+    ),
+)
+def test_find_small_files(
+    files: List[Dict[str, int]],
+    user_pages: int,
+    split_multipage_doc_setup: str,
+    expected_output,
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1540]: typing, please

File: annotation/tests/test_distribution.py (hunk 29 above)
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1378]: Im sorry about that. My intention was to cover when user_id is not in example_db with the second test case. I guess in the meantime I copy-pasted the lists from above.

File: annotation/tests/test_distribution.py (hunk 1 above)
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1326]: Mocked the SPLIT_MULTIPAGE_DOC var and simplified the test

File: annotation/tests/test_distribution.py (hunk 20 above)
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1480]: Sorry again. Now I know what you meant by simplify the parametrize

File: annotation/annotation/models.py [hunk 32]
```diff
@@ -200,6 +200,17 @@ class User(Base):
...
     )
 
+    def __eq__(self: "User", other: Any) -> bool:
+        if not isinstance(other, User):
+            return NotImplemented
+
+        for column in self.__table__.columns:
+            if column.key != "_sa_instance_state" and other.__getattribute__(
+                column.key
+            ) != self.__getattribute__(column.key):
```
Comment (cakeinsauce) [inline annotation/annotation/models.py:210]:
```suggestion
            ) != other.__getattribute__(column.key):
```

File: annotation/tests/test_distribution.py [hunk 33]
```diff
@@ -1277,3 +1290,285 @@ def test_get_page_number_combinations(
...
+            "",
+            ValidationSchema.cross,
+            (0, 1),
+            [],
+        ),
+        (
+            "1",
+            ValidationSchema.hierarchical,
+            1,
+            [2],
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1313]: list -> tuple

File: annotation/annotation/models.py (hunk 32 above)
Comment (UkiStefanovic) [reply annotation/annotation/models.py:210]: How did I not notice that... Sorry

File: annotation/tests/test_distribution.py [hunk 34]
```diff
@@ -1277,3 +1290,285 @@ def test_get_page_number_combinations(
...
+    files_page_count: List[int],
+    user_pages: int,
+    expected_output: List[Dict[str, int]],
+):
+    file_count = len(files_page_count)
+    files = [
+        {"pages_number": x, "file_id": i}
+        for x, i in zip(files_page_count, range(file_count))
+    ]
+    print(files)
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1511]: remove print please

File: annotation/tests/test_distribution.py (hunk 20 above)
Comment (cakeinsauce) [reply annotation/tests/test_distribution.py]: it is still the same
image

File: annotation/tests/test_distribution.py (hunk 21 above)
Comment (cakeinsauce) [reply annotation/tests/test_distribution.py]: still the same
image

File: annotation/tests/test_distribution.py [hunk 35]
```diff
@@ -1277,3 +1290,285 @@ def test_get_page_number_combinations(
...
+        split_multipage_doc_setup,
+    ):
+        output = choose_validators_users(
+            validation_type,
+            annotators,
+            validators,
+            annotation_tasks=(
+                [{"user_id": user_id} for user_id in annotation_tasks_ids]
+                if isinstance(annotation_tasks_ids, Iterable)
+                else [{"user_id": annotation_tasks_ids}]
```
[the remark this answers is not in the document]
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1337]: Yeah, I've only now noticed that (1) is an int, but (1,) is a tuple with one element

File: annotation/annotation/models.py [hunk 36]
```diff
@@ -200,6 +200,17 @@ class User(Base):
         "Job", secondary=association_job_owner, back_populates="owners"
     )
 
+    def __eq__(self: "User", other: object) -> bool:
```
Comment (cakeinsauce) [inline annotation/annotation/models.py:203]: never type self, excessive

File: annotation/tests/test_distribution.py (hunk 20 above)
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py]: ⁠Regarding this, I thought that i cant remove file_id's. If we have multiple files with the same page_number, then we should keep file_id's to check if the output was as expected. But if i want to just see if the tests work correctly, i could just:

`​assert sum([file["pages_number"] for file in output]) == user_pages`

In that case, file_id's are redundant and I'll remove them completely

Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py]: Nevermind that, I think I got it

File: annotation/annotation/models.py (hunk 36 above)
Comment (cakeinsauce) [reply annotation/annotation/models.py:203]: you have to have self in any method, but you shouldn't specify typing for self, it is obvious the same as the class in which this method is located

**Review by cakeinsauce — APPROVED** [review]
Great job!
